// @antv/g6/lib/behavior/drag-node.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/*
 * @Author: moyee
 * @Date: 2019-06-27 18:12:06
 * @LastEditors: moyee
 * @LastEditTime: 2019-08-22 18:41:45
 * @Description: 拖动节点的Behavior
 */
var _require = require('lodash'),
    merge = _require.merge,
    isString = _require.isString;

var _require2 = require('../global'),
    delegateStyle = _require2.delegateStyle;

var body = document.body;
module.exports = {
  getDefaultCfg: function getDefaultCfg() {
    return {
      updateEdge: true,
      delegateStyle: {}
    };
  },
  getEvents: function getEvents() {
    return {
      'node:dragstart': 'onDragStart',
      'node:drag': 'onDrag',
      'node:dragend': 'onDragEnd',
      'canvas:mouseleave': 'onOutOfRange'
    };
  },
  onDragStart: function onDragStart(e) {
    var _this = this;

    if (!this.shouldBegin.call(this, e)) {
      return;
    }

    var item = e.item;
    var graph = this.graph;
    this.targets = []; // 获取所有选中的元素

    var nodes = graph.findAllByState('node', 'selected');
    var currentNodeId = item.get('id'); // 当前拖动的节点是否是选中的节点

    var dragNodes = nodes.filter(function (node) {
      var nodeId = node.get('id');
      return currentNodeId === nodeId;
    }); // 只拖动当前节点

    if (dragNodes.length === 0) {
      this.target = item;
    } else {
      // 拖动多个节点
      if (nodes.length > 1) {
        nodes.forEach(function (node) {
          _this.targets.push(node);
        });
      } else {
        this.targets.push(item);
      }
    }

    this.origin = {
      x: e.x,
      y: e.y
    };
    this.point = {};
    this.originPoint = {};
  },
  onDrag: function onDrag(e) {
    if (!this.origin) {
      return;
    }

    if (!this.get('shouldUpdate').call(this, e)) {
      return;
    } // 当targets中元素时，则说明拖动的是多个选中的元素


    if (this.targets.length > 0) {
      this._updateDelegate(e);
    } else {
      // 只拖动单个元素
      this._update(this.target, e, true);
    }
  },
  onDragEnd: function onDragEnd(e) {
    var _this2 = this;

    if (!this.origin || !this.shouldEnd.call(this, e)) {
      return;
    }

    if (this.shape) {
      this.shape.remove();
      this.shape = null;
    }

    if (this.target) {
      var delegateShape = this.target.get('delegateShape');

      if (delegateShape) {
        delegateShape.remove();
        this.target.set('delegateShape', null);
      }
    }

    if (this.targets.length > 0) {
      // 获取所有已经选中的节点
      this.targets.forEach(function (node) {
        return _this2._update(node, e);
      });
    } else if (this.target) {
      this._update(this.target, e);
    }

    this.point = {};
    this.origin = null;
    this.originPoint = {};
    this.targets.length = 0;
    this.target = null; // 终止时需要判断此时是否在监听画布外的 mouseup 事件，若有则解绑

    var fn = this.fn;

    if (fn) {
      body.removeEventListener('mouseup', fn, false);
      this.fn = null;
    }
  },
  // 若在拖拽时，鼠标移出画布区域，此时放开鼠标无法终止 drag 行为。在画布外监听 mouseup 事件，放开则终止
  onOutOfRange: function onOutOfRange(e) {
    var self = this;

    if (this.origin) {
      var canvasElement = self.graph.get('canvas').get('el');

      var fn = function fn(ev) {
        if (ev.target !== canvasElement) {
          self.onDragEnd(e);
        }
      };

      this.fn = fn;
      body.addEventListener('mouseup', fn, false);
    }
  },
  _update: function _update(item, e, force) {
    var origin = this.origin;
    var model = item.get('model');
    var nodeId = item.get('id');

    if (!this.point[nodeId]) {
      this.point[nodeId] = {
        x: model.x,
        y: model.y
      };
    }

    var x = e.x - origin.x + this.point[nodeId].x;
    var y = e.y - origin.y + this.point[nodeId].y; // 拖动单个未选中元素

    if (force) {
      this._updateDelegate(e, x, y);

      return;
    }

    var pos = {
      x: x,
      y: y
    };

    if (this.get('updateEdge')) {
      this.graph.updateItem(item, pos);
    } else {
      item.updatePosition(pos);
      this.graph.paint();
    }
  },

  /**
   * 更新拖动元素时的delegate
   * @param {Event} e 事件句柄
   * @param {number} x 拖动单个元素时候的x坐标
   * @param {number} y 拖动单个元素时候的y坐标
   */
  _updateDelegate: function _updateDelegate(e, x, y) {
    var bbox = e.item.get('keyShape').getBBox();

    if (!this.shape) {
      // 拖动多个
      var parent = this.graph.get('group');
      var attrs = merge({}, delegateStyle, this.delegateStyle);

      if (this.targets.length > 0) {
        var _this$calculationGrou = this.calculationGroupPosition(),
            _x = _this$calculationGrou.x,
            _y = _this$calculationGrou.y,
            width = _this$calculationGrou.width,
            height = _this$calculationGrou.height,
            minX = _this$calculationGrou.minX,
            minY = _this$calculationGrou.minY;

        this.originPoint = {
          x: _x,
          y: _y,
          width: width,
          height: height,
          minX: minX,
          minY: minY
        }; // model上的x, y是相对于图形中心的，delegateShape是g实例，x,y是绝对坐标

        this.shape = parent.addShape('rect', {
          attrs: _extends({
            width: width,
            height: height,
            x: _x,
            y: _y
          }, attrs)
        });
      } else if (this.target) {
        this.shape = parent.addShape('rect', {
          attrs: _extends({
            width: bbox.width,
            height: bbox.height,
            x: x - bbox.width / 2,
            y: y - bbox.height / 2
          }, attrs)
        });
        this.target.set('delegateShape', this.shape);
      }

      this.shape.set('capture', false);
    }

    if (this.targets.length > 0) {
      var clientX = e.x - this.origin.x + this.originPoint.minX;
      var clientY = e.y - this.origin.y + this.originPoint.minY;
      this.shape.attr({
        x: clientX,
        y: clientY
      });
    } else if (this.target) {
      this.shape.attr({
        x: x - bbox.width / 2,
        y: y - bbox.height / 2
      });
    }

    this.graph.paint();
  },

  /**
   * 计算delegate位置，包括左上角左边及宽度和高度
   * @memberof ItemGroup
   * @return {object} 计算出来的delegate坐标信息及宽高
   */
  calculationGroupPosition: function calculationGroupPosition() {
    var graph = this.graph;
    var nodes = graph.findAllByState('node', 'selected');
    var minx = [];
    var maxx = [];
    var miny = [];
    var maxy = []; // 获取已节点的所有最大最小x y值

    for (var _iterator = nodes, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref;

      if (_isArray) {
        if (_i >= _iterator.length) break;
        _ref = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) break;
        _ref = _i.value;
      }

      var id = _ref;
      var element = isString(id) ? graph.findById(id) : id;
      var bbox = element.getBBox();
      var _minX = bbox.minX,
          _minY = bbox.minY,
          _maxX = bbox.maxX,
          _maxY = bbox.maxY;
      minx.push(_minX);
      miny.push(_minY);
      maxx.push(_maxX);
      maxy.push(_maxY);
    } // 从上一步获取的数组中，筛选出最小和最大值


    var minX = Math.floor(Math.min.apply(Math, minx));
    var maxX = Math.floor(Math.max.apply(Math, maxx));
    var minY = Math.floor(Math.min.apply(Math, miny));
    var maxY = Math.floor(Math.max.apply(Math, maxy));
    var x = minX - 20;
    var y = minY + 10;
    var width = maxX - minX;
    var height = maxY - minY;
    return {
      x: x,
      y: y,
      width: width,
      height: height,
      minX: minX,
      minY: minY
    };
  }
};
